File: codeX-frontend/src/services/native/terminal.js
/**
 * Native Terminal Service
 * Communicates with the real local PTY engine on desktop.
 * Provides clean multi-session management, streaming I/O, and resizing.
 */

import { isDesktopApp } from './platform.js';

class NativeTerminalService {
  constructor() {
    this.activeListeners = new Map();
  }

  isSupported() {
    return isDesktopApp();
  }

  async getShells() {
    if (!isDesktopApp()) {
      return [{ id: 'web-term', name: 'Web Terminal', path: 'web', is_default: true }];
    }
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('term_get_shells');
  }

  async getDefaultShell() {
    if (!isDesktopApp()) return 'web';
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('term_get_default_shell');
  }

  async createSession({ id, title, shell, cwd, cols = 80, rows = 24, onData, onExit }) {
    if (!isDesktopApp()) {
      throw new Error('Local PTY terminal is only supported in CodeX desktop application');
    }

    const { invoke } = await import('@tauri-apps/api/core');
    const { listen } = await import('@tauri-apps/api/event');

    const sessionInfo = await invoke('term_create', {
      id,
      title: title || null,
      shell: shell || null,
      cwd: cwd || null,
      cols,
      rows,
    });

    // Listen for PTY stdout/stderr stream
    const unlistenData = await listen(`terminal-output-${id}`, (event) => {
      if (onData && typeof event.payload === 'string') {
        onData(event.payload);
      }
    });

    // Listen for terminal process exit
    const unlistenExit = await listen(`terminal-exit-${id}`, (event) => {
      if (onExit) {
        onExit(event.payload);
      }
      this.cleanupListeners(id);
    });

    this.activeListeners.set(id, [unlistenData, unlistenExit]);

    return {
      id,
      title: sessionInfo.title,
      shell: sessionInfo.shell,
      cwd: sessionInfo.cwd,
      write: (data) => this.write(id, data),
      resize: (newCols, newRows) => this.resize(id, newCols, newRows),
      close: () => this.close(id),
    };
  }

  async write(id, data) {
    if (!isDesktopApp()) return;
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('term_write', { id, data });
  }

  async resize(id, cols, rows) {
    if (!isDesktopApp()) return;
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('term_resize', { id, cols, rows });
  }

  async close(id) {
    this.cleanupListeners(id);
    if (!isDesktopApp()) return;
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('term_close', { id });
  }

  async listSessions() {
    if (!isDesktopApp()) return [];
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('term_list');
  }

  cleanupListeners(id) {
    if (this.activeListeners.has(id)) {
      const unlisteners = this.activeListeners.get(id);
      unlisteners.forEach((fn) => {
        try {
          fn();
        } catch {}
      });
      this.activeListeners.delete(id);
    }
  }
}

export const terminalService = new NativeTerminalService();

File: codeX-frontend/src-tauri/src/commands/terminal.rs
use crate::platform::{self, ShellInfo};
use crate::terminal::{TerminalManager, TerminalSessionInfo};
use tauri::{AppHandle, State};

#[tauri::command]
pub fn term_get_shells() -> Vec<ShellInfo> {
    platform::get_available_shells()
}

#[tauri::command]
pub fn term_get_default_shell() -> String {
    platform::get_default_shell()
}

#[tauri::command]
pub fn term_create(
    app: AppHandle,
    term_state: State<'_, TerminalManager>,
    id: String,
    title: Option<String>,
    shell: Option<String>,
    cwd: Option<String>,
    cols: Option<u16>,
    rows: Option<u16>,
) -> Result<TerminalSessionInfo, String> {
    let chosen_shell = shell.unwrap_or_else(platform::get_default_shell);
    let chosen_cwd = cwd.unwrap_or_else(|| {
        std::env::var("HOME").unwrap_or_else(|_| ".".to_string())
    });

    term_state.create_session(
        app,
        id,
        title,
        &chosen_shell,
        &chosen_cwd,
        cols.unwrap_or(80),
        rows.unwrap_or(24),
    )
}

#[tauri::command]
pub fn term_write(
    term_state: State<'_, TerminalManager>,
    id: String,
    data: String,
) -> Result<(), String> {
    term_state.write_session(&id, data.as_bytes())
}

#[tauri::command]
pub fn term_resize(
    term_state: State<'_, TerminalManager>,
    id: String,
    cols: u16,
    rows: u16,
) -> Result<(), String> {
    term_state.resize_session(&id, cols, rows)
}

#[tauri::command]
pub fn term_close(term_state: State<'_, TerminalManager>, id: String) -> Result<(), String> {
    term_state.close_session(&id)
}

#[tauri::command]
pub fn term_rename(
    term_state: State<'_, TerminalManager>,
    id: String,
    title: String,
) -> Result<(), String> {
    term_state.rename_session(&id, title)
}

#[tauri::command]
pub fn term_list(term_state: State<'_, TerminalManager>) -> Vec<TerminalSessionInfo> {
    term_state.list_sessions()
}

